finalfrontier/src/vocab/subword.rs

```diff
@@ -196,6 +200,51 @@ where
     }
 }
 
+/// Constructs a `SubwordVocab` from a `VocabBuilder<T>` where `T: Into<String>`.
+impl<T> From<VocabBuilder<SubwordVocabConfig<NGramConfig>, T>>
+    for SubwordVocab<NGramConfig, NGramIndexer>
+where
+    T: Hash + Eq + Into<String>,
+{
+    fn from(builder: VocabBuilder<SubwordVocabConfig<NGramConfig>, T>) -> Self {
+        let config = builder.config;
+        let mut words = Vec::new();
+        let mut ngram_counts = HashMap::new();
+        for (word, count) in builder.items {
+            let word = word.into();
+            if word != util::EOS && count < config.min_count as usize {
+                continue;
+            }
+            let word = Word::new(word, count);
+            for ngram in NGrams::new(
+                &bracket(word.label()),
+                config.min_n as usize,
+                config.max_n as usize,
+            )
+            .map(|ngram| ngram.to_string())
+            {
+                let cnt = ngram_counts.entry(ngram).or_insert(0);
+                *cnt += count;
+            }
+            words.push(word);
+        }
+
+        let mut ngrams = ngram_counts
+            .into_iter()
+            .filter(|(_, count)| *count >= config.indexer.min_ngram_count as usize)
+            .map(|(ngram, count)| CountedType::new(ngram, count))
+            .collect::<Vec<_>>();
+
+        words.sort_unstable_by(|w1, w2| w2.cmp(&w1));
+        ngrams.sort_unstable_by(|w1, w2| w2.cmp(&w1));
+        let ngrams = ngrams
+            .into_iter()
+            .map(|ngram| ngram.label)
+            .collect::<Vec<_>>();
```

Comment:
ngrams go from a map to a vec to another vec. What is the memory impact?

Reply:
I know it's not ideal, but how else could we handle this?

We want to keep only ngrams coming from in-vocab words, others wouldn't be trained. Therefore we need to extract the ngrams when we construct the vocab.

We want to filter ngrams by a distinct `min_count`, so we need to count how often the ngram appears in the in-vocab words which means we have to put them into a map (note, that's the same we do for the words, just at a different time.).

We also want to sort the ngrams by frequency, which means we need some sortable collection, `Vec` being the most straight forward choice here. 

`finalfusion::NGramIndexer` stores ngrams as `String`s, so we need to unwrap the `CountedType`s `label`.

What do you propose here?

Reply:
Fwiw, training on twe with `mincount=30` takes up ~4.6gb RAM. Vocab construction also didn't lead to a spike in required memory.

Training with the same parameters with a bucket vocab requries ~4.4gb RAM.

Memory usage is dominated by the embedding matrices. With a lower `mincount` for ngrams, the size of the subword-matrix explodes and we end up with ~20gb memory consumption during training. But that's due to the large amount of ngram-embeddings and not due to the number of stored ngrams.

Reply:
At least, it is trivial to go from 3|V| to 2|V|: after constructing the `HashMap`, create the final `Vec` of n-grams (without counts) from the keys of the `HashMap`, retaining the keys of n-grams that occur `>= mincount`. Then sort this `Vec` using the `HashMap` (so, when comparing two n-grams, you look them up in the hashmap rather than storing the count in the `Vec` with the n-gram).

Reply:
Wouldn't `ngram_counts.into_iter()` consume the `HashMap` and free the memory, and wouldn't the same hold for unwrapping collecting the final `Vec<String>` from the intermediate `Vec<CountedType<String>>`?

Reply:
I don't know. The `Vec` would become owned of the `String` heap objects, but I don't know if `into_iter`  keeps the `HashMap` itself or whether it incrementally deallocates it. Maybe look at the implementations.

Reply:
`Vec` seems to keep the memory for the `Vec`, but of course, the heap-allocated `String`s are moved to the new owner. Since most of the overhead is probably in the `String`s and not in the 'packed' counts, I am fine with this as-is.

Reply:
I have already pushed the version that you suggested. Do you want me to revert? I doubt either version will be a bottleneck in any way.

Reply:
Choose the version that you think has the best readability :).

Reply:
Let's stick with this one.